### ISSUE TITLE
Annex mode siblings can't deal with credential change

### ISSUE THREAD

**Issue by mslw** [datalad/datalad-next#59]
Context: git-annex caches the username & password in read-only plaintext file `.git/annex/creds/<annex-uuid>`. I wanted to check what would happen if DataLad credentials got updated (whether credentials coming from DataLad take precedence).

What I did:
- created two siblings (dav1 using filetree mode, dav2 using export mode) on Sciebo using `create_sibling_webdav`
- removed Sciebo token used previously, and generated a new one
- updated the credentials stored by DataLad with `datalad credentials set (...) :secret`

What I observed:
- Cloning / getting from either sibling works.
- Pushing from the original dataset to the filetree sibling works
- Pushing from the original dataset to the annex sibling fails with the following error

```
Push to 'dav2':  25%|████████████████████▎                                                            | 1.00/4.00 [00:00<00:00, 10.3k Steps/sCommandError: 'git -c diff.ignoreSubmodules=none annex copy --batch -z --to dav2-storage --fast --json --json-error-messages --json-progress -c annex.dotfiles=true' failed with exitcode 1 under /home/mszczepanik/Documents/test-next/example [info keys: stdout_json]                     
> to dav2-storage...                                                                                                                          
  DAV failure: Status {statusCode = 401, statusMessage = "Unauthorized"} "<?xml version=\"1.0\" encoding=\"utf-8\"?>\n<d:error xmlns:d=\"DAV:\" xmlns:s=\"http://sabredav.org/ns\">\n  <s:exception>Sabre\\DAV\\Exception\\NotAuthenticated</s:exception>\n  <s:message>No public access to this resource., Username or password was incorrect, No 'Authorization: Bearer' header found. Either the client didn't send one, or the server is mis-configured, Username or password was incorrect</s:message>\n</d:error>\n" HTTP request: "PUT" "/remote.php/dav/files/m.szczepanik%40fz-juelich.de/webdav-walkthrough/annex-mode/git-annex-webdav-tmp-MD5E-s7576199--5c6bc89919f5e45cc9e7081fa3aac472.jpg"
  DAV failure: Status {statusCode = 401, statusMessage = "Unauthorized"} "<?xml version=\"1.0\" encoding=\"utf-8\"?>\n<d:error xmlns:d=\"DAV:\" xmlns:s=\"http://sabredav.org/ns\">\n  <s:exception>Sabre\\DAV\\Exception\\NotAuthenticated</s:exception>\n  <s:message>No public access to this resource., Username or password was incorrect, No 'Authorization: Bearer' header found. Either the client didn't send one, or the server is mis-configured, Username or password was incorrect</s:message>\n</d:error>\n" HTTP request: "PUT" "/remote.php/dav/files/m.szczepanik%40fz-juelich.de/webdav-walkthrough/annex-mode/git-annex-webdav-tmp-MD5E-s7576199--5c6bc89919f5e45cc9e7081fa3aac472.jpg"
  This could have failed because --fast is enabled.
copy: 1 failed
```
 
I am wondering what could be the cause, given that the exporttree seems to be the only difference.

**Comment by mslw** [comment]
That wouldn't be my immediate course of action - what's good for Annex should be OK for us.

In this case it would be a good idea to understand what causes the issue (at which stage of push it happens), and I didn't follow back then. But it is weird - looks like the DataLad credentials do take precedence for the filetree sibling, but not annex sibling?

Do you know what "`--fast` is enabled" means in this context?

**Comment by mih** [comment]
`--fast` is from `git annex copy` and does

```
      --fast When copying content to a remote, avoid a round trip to check if the remote already has content.
              This can be faster, but might skip copying content to the remote in some cases.
```

we add it whenever `push --force` is not `all` or `checkdatapresent`.

This is indeed something that differs between `push` to a normal annex or an export.

And the other thing is the credential setup. We only have and use new-style credentials for the push-to-export. For the traditional `copy` we use the code in the core package, which does not support credential deployment at all, hence relies on the outdated credential copy in the repo.

So basically, this issue should be transferred to the core project, and needs to wait there till it can do credential deployment.

**Comment by mih** [comment]
I just found the config setting `annex.cachecreds`

> When  "true"  (the  default), git-annex will cache credentials used to access special remotes in files in .git/annex/creds/ that only you can read. To disable that caching, set to "false",  and credentials will  only be read from the environment, or if they have been embedded in encrypted form in the git repository, will be extracted and decrypted each time git-annex needs to  access the remote.

It sounds as if we could enable this setting by default, whenever we are dealing with a scenario where datalad deploys the credentials (in the environment), like for webdav.
